### ISSUE TITLE
Scope of the crypto module

### ISSUE THREAD

**Issue by asmeurer** [sympy/sympy#18269]
I think we should outline a scope for the crypto submodule, to clarify what sorts of things should and should not go in the package. 

Currently the [documentation of the module says](https://docs.sympy.org/1.5.1/modules/crypto.html) says

> **Warning** This module is intended for educational purposes only. Do not use the functions in this module for real cryptographic applications. If you wish to encrypt real data, we recommend using something like the cryptography module.

I've also stated my opinion on the [mailing list](https://groups.google.com/d/msg/sympy/7XMYGWk6vqM/U7PHoFpRAQAJ) that the for GSoC projects I'd prefer to see things in the crypto module that are useful outside of just crypto, but this doesn't necessarily need to be part of the general scope.

My main concern is that I don't want people to use the crypto module for actual cryptography, even despite the above warning. So I don't think we should implement any sort of "modern" cryptographic primitive.

What would be a good way to delimit what should and should not go in the module?

**Comment by namannimmo10** [comment]
A lot of simple ciphers (as mentioned [here](http://www.crypto-it.net/eng/simple/index.html)) have not been yet implemented in the crypto module. I still haven't picked up any topic specifically for GSoC this year. But the ` crypto` module sounds interesting. 
> I'd prefer to see things in the crypto module that are useful outside of just crypto,

Can you please suggest any ideas I can work upon?

**Comment by sylee957** [comment]
I'm not very clear that things that are too much 'modern' can be against 'educational' purpose. 

I think that 'educational purpose' should refer to things that have clean mathematical model, and it may or may not be related to the historical year that the cipher was invented.

**Comment by asmeurer** [comment]
It's not that being modern is uneducational. It's that I don't want people to use the code in SymPy for actual cryptographic applications.

**Comment by sylee957** [comment]
I don't think that any people who works on actual cryptographic applications would use this library, but we should try to have things that are useful in some other ways than some direct applications, like introducing a easy way to study the modern cryptography, and this can also be corresponding to the educational purpose as well.

Since this is a symbolic math library, we can be focused on things related to the symbolics.
In this sense, I think the real problem is that most of the stuff in `crypto` module are still too much numeric and is not contributing to the symbolic computations.

So at some point, we should try not to introduce stuff that operates on python lists or strings and do some computations, but try to introduce some stuff that can be generalized:

- A symbolic concept for string, alphabet sets can be introduced.
- Make most of the ciphers to operate on canonical integers instead of alphabets or strings, and combine with the concept above to make things like `rot-13` non-trivial.
- Make use of the features like `Permutation`, and contribute back some ideas back to the combinatorics module, like some ideas of implicit or sparse permutations.
- Introduce bit-wise arithmetic, formula manipulations, equation solving.
- Introduce number-theoretic formula manipulations and equation solving.
- Make ciphers `Basic` and introduce compositions, inverses, transformations of the ciphers
- Finally, connect with the practical cryptography by code generations and lambdify.

**Comment by smichr** [comment]
Whatever we can do to keep this from being a block hole that collects historically interesting, but provenly weak methods of encrypting text would be good. Thanks for the suggestions, @sylee957.

**Comment by czgdp1807** [comment]
I was wondering that if we can check how the implementation in `crypto` module responds to symbolic inputs? For example is there a possibility to get an encrypted message with a symbolic key and then substituting that key with various numerical values.

**Comment by asmeurer** [comment]
I think for at least some of the algorithms that would require some sort of symbolic string manipulations. But others may not, e.g., RSA basically works on numbers. I think doing things symbolically not only avoids a lot of the issues I am concerned about, but it also makes things a lot better for education, as you can actually see what the cryptosystem is doing beyond just reading the code.

**Comment by smichr** [comment]
Here's an example of how [this page](https://www.cryptool.org/en/cto-ciphers/autokey) describes a cipher:
```
Let s be such a reversible function. Then the Vigenère encryption for an input character in and a key key can be described as:

out = s^-1(s(in) + s(key))
```

**Comment by sylee957** [comment]
Interesting. I think that it can be always able to split every cryptography into parts, such that
- Every ciphers should operate on integers
- The mapping between strings and integers are trivial, and can be made configurable.

But I'd have a question whether this can be generalized to describe every cryptographic means used in the world, by introducing some additional concepts like vectors or matrices.

**Comment by asmeurer** [comment]
> Whatever we can do to keep this from being a block hole that collects historically interesting, but provenly weak methods of encrypting text would be good. Thanks for the suggestions, @sylee957.

Wasn't that the original purpose of the module? It seems like most of what was there from the start were historically interesting but practically useless methods, like Vigenère or Caesar ciphers. 

I personally am more worried about the opposite: more "modern" cyphers that seem like they are useful, but they actually shouldn't be used because the SymPy implementations are not suitable for practical usage. For example, [we don't use secure random number generation](https://github.com/sympy/sympy/blob/1a66b86d23cfc606ad3ad4185aaec40ee9b822a5/sympy/crypto/crypto.py#L33), which cause major issues for algorithms that rely on that. Or see for instance https://www.johndcook.com/blog/2019/02/11/rsa-duplication-flaws/, which discusses how making RSA secure is much more subtle than the mathematics would suggest. I'm not as worried about ciphers that are so easy to break that they appear as [puzzles in the daily newspaper](https://en.wikipedia.org/wiki/Cryptogram). 

If someone wants to use SymPy's implementation of ROT13 to do real ROT13, that's no big deal, because ROT13 is only ever used for jokes and Easter eggs. But if they try to use SymPy's RSA to do real encryption, that's bad.

It sounds like between the two of us, we aren't really happy with anything in this module. But I do think that having purely symbolic representations is the way forward.

**Comment by sylee957** [comment]
In my opinion, RSA is just about a special computation of modular exponentiation, that bijectively maps an integer to an another integer, and it has nothing to deal with 'security' in mathematical sense.

I wonder why should we be so serious about security, when introducing something in terms of modern cryptography. 
I think that if the computation is done right, then it will be fine. 
It's up to the responsibility of users who are going to use it in some application or not. But it's not our responsibility to make it secure or not.

**Comment by asmeurer** [comment]
I strongly disagree with that sentiment. If that's the case, then why is the module called "crypto"? If RSA is just some group operations, then we should be building out the group theory module to represent what it does generically. From that point of view, RSA is just a certain calculation on groups that doesn't even need to be written down as a separate library function. Cryptosystems like RSA only matter from the point of view of cryptography, because they transform computationally hard problems like factoring or discrete logarithms into an asymmetric encryption scheme.

The reason this concerns me is that I've seen numerous examples where people use bad cryptography. That blog post I linked to shows some examples. I don't want SymPy to have any part in that, because it happens to have a submodule labeled "crypto". I also find most of the code in the current crypto module to be out of place for SymPy, since it isn't symbolic in nature and makes limited use of the rest of SymPy. 

I agree we *shouldn't* be concerned with security. But the fact remains that if something is there, people will use it. So if our system isn't practically secure by design, we shouldn't even make it look like something that is. I believe having it be completely symbolic would achieve this. I also think removing it entirely would do the same, but that's obviously a more aggressive move. As it is now, it isn't really clear what the point of the module is, if you aren't actually supposed to use it. I guess the point is "educational", meaning it should be instructive to people to read the source code? What is the point of it beyond that?

**Comment by smichr** [comment]
I can find the original PR #1854 but I can't find the discussion that I recall regarding the desire to have this here. I recall that it was educationally motivated in a situation where SymPy was already being used for some things and it was desired to include the crypto, too. But the beauty of python modules is that you import the modules and they don't need to know about one another...so you can study with SymPy and your favorite crypto library at the same time.

There is a good tutorial [here](https://www.tutorialspoint.com/cryptography_with_python/cryptography_with_python_quick_guide.htm) that would be good for educational purposes.

There is a library for doing serious crypto [here](https://github.com/pyca/cryptography)

And a library that has many classical ciphers is [here](https://github.com/jameslyons/pycipher/tree/master/pycipher).

I am in favor of removing this module. I'm not sure it is worth time figuring out how to make it represent the ciphers mathematically.

**Comment by asmeurer** [comment]
The original discussion was on the mailing list https://groups.google.com/d/msg/sympy/oU6sDDtJpVw/1jnb14tuOyYJ. It doesn't look like any of these concerns were discussed at the time. 

@wdjoyner, you may want to weigh in on this discussion.

**Comment by wdjoyner** [comment]
I saw the code written by James Lyons references above. He wrote an email to me years ago that his python code is licensed using the "MIT license for all the code." 

I'm not using my code any longer and won't be offended if you want to remove mine and use his instead. I personally found my interface simpler but I wrote it, so am biased:-) 
BTW, (a) I'm retired now and no longer use the code (though there are lecture notes out there which reference it, but I don't know if they're used either), (b) Sage is much easier to install and use on windows now, to the need I had for crypto code in Sympy no longer applies.

**Comment by sylee957** [comment]
I had a different point of view in terms of what is 'educational'.

As I had made a extension of RSA to work with any number of primes, there can be more esoteric extensions of RSA that are not implemented in any practical crypto library, and not many popular materials are even aware about.

So, the point I'd talk about is, similarly as RSA as I have noted, the modern cryptography can become a more difficult math exercise than how it is used in real applications, and some students may need the references and see some papers implemented as code.

I'd agree on moving this somewhere else, like `combinatorics` module, to focus more on solving the puzzles. But I'd wonder the naming of the module was so significant that it could confuse people who are already professional enough to write the applications.

**Comment by smichr** [comment]
> some students may need the references and see some papers implemented as code.

The CodePoint link I gave above (see "good tutorial") is good about this. I agree that seeing the code is helpful. I like the idea of putting it into combinatoric. We have the Rubik's cube there. Much if ciphering has to do with transformations on a 2D matrix or a 1D line with modular addition.

**Comment by sylee957** [comment]
Okay, I'll push the idea of making `crypto` a stub page and moving everything to `combinatorics`.
If @asmeurer doesn't like the original warning statement, it can also be modified like

- **Warning** This module is intended for educational purposes only. And should also be focused about studying the math of the cryptography rather than studying the criterion about the actual security. ...

And yes, the education would also not mean by teaching people about what is secure key generation or such, but whether it's a possible idea to pass and recover the integers in some weird formulas.

**Comment by czgdp1807** [comment]
> Okay, I'll push the idea of making `crypto` a stub page and moving everything to `combinatorics`.

What should be done with currently open PRs for `crypto` module? Should they be stalled, or we should ask the contributors to move their code to `combinatorics` module from where it can directly go to the `combinatorics` module of `master`.

**Comment by asmeurer** [comment]
I don't think moving this to another submodule solves all of the concerns here. 

> If @asmeurer doesn't like the original warning statement, it can also be modified like

I don't have a problem with the warning statement (after all, [I wrote it](https://github.com/sympy/sympy/pull/13687)). I'm questioning whether the module, or at least which parts of it, should be in scope for SymPy in the first place. 

> What should be done with currently open PRs for crypto module? Should they be stalled, or we should ask the contributors to move their code to combinatorics module from where it can directly go to the combinatorics module of master.

I'm not sure if we've reached a consensus here yet. Should we stall crypto PRs? I suppose if we end up deciding to remove the code that would be added in a new PR, it wouldn't matter if it were already merged. 

I will say that I would personally rather not see any GSoC projects for the crypto module (though this should also be discussed with the community).

**Comment by asmeurer** [comment]
Aside from the security concerns, I do also question how this module fits into SymPy in the first place. It doesn't work symbolically at all. This isn't necessarily bad—for instance, the combinatorics module mostly doesn't work symbolically—however unlike combinatorics nothing in this module would ever be useful for other modules. We have many such "leaf" submodules in SymPy, so that doesn't necessarily rule it out. My main reason for picking on `crypto` is 1, the security concerns, and 2, I see it becoming a popular thing for people to add things to, and like @smichr I don't think it would be very helpful for it to become a dumping grounds.

But most importantly, I think we should reach a consensus here, so that everyone is happy with whatever we decide to do.

**Comment by oscarbenjamin** [comment]
I don't see why crypto should be part of sympy. It could be a separate project.

I ignore all PRs for crypto because I don't want to spend time learning the module myself. Looking at it quickly now though I think that it wouldn't be hard to remove any dependency on sympy at all and make it a totally independent project.

**Comment by sylee957** [comment]
I agree that lots of recent contributions of crypto module are lacking some depth and some people may exploit such stuff for GSoC opportunity.

But I actually think that seeing some new people coming up with the ideas is not a totally bad sign. Because it can also be an indicator of how people acknowledge sympy for.

I'm still thinking of how it can be made better than kicking out of, because cryptography can still be a valid addition to symbolic math as how Sage does
http://doc.sagemath.org/html/en/reference/cryptography/index.html
But right now, I agree that there aren't much skeleton worked on, like symbolic string or bitwise arithmetics.
I wonder if the module can come back in more better shape, in the future.

**Comment by czgdp1807** [comment]
IMO, we should keep `crypto` module. Wolfram too has Cryptography. See, https://www.wolfram.com/language/12/cryptography/receive-encrypted-messages.html?product=language
Though it should be made more symbolic, like `subs` should work on intermediate encrypted messages. `f(message, key)` should be returned and on using `.subs(key, some_value)`, the encrypted message should be returned.

**Comment by smichr** [comment]
There is a [short dissertation](https://ttu-ir.tdl.org/bitstream/handle/2346/46988/RAY-THESIS.pdf) about the Chao Cipher that shows how a certain cipher acts as Permutations on `Z26`. Maybe there could be some ideas for how to make the symbolic nature of text transformations more compatible with the algebraic nature of SymPy.

> f(message, key) should be returned and on using .subs(key, some_value)

I'm not sure this works real well with transformations in general. There may be several parameters besides the key that determine how a transformation should work. e.g. [ScrabbleCipher](https://prgomez.com/scrabble-cipher/) and the javascript that he gives for it.

**Comment by czgdp1807** [comment]
Well, I just gave this another thought. IMHO, `crypto` can be removed. There is an awesome project on cryptography. See, https://pypi.org/project/cryptography/
Is there any to find out how many users actually use `crypto` module? Can we use the number of hits on the documentation of `sympy.crypto` web page. If not many people are there using it then we can proceed deprecating it.

**Comment by wdjoyner** [comment]
On Wed, Mar 18, 2020 at 3:12 PM Gagandeep Singh <notifications@github.com>
wrote:

> Well, I just gave this another thought. IMHO, crypto can be removed.
> There is an awesome project on cryptography. See,
> https://pypi.org/project/cryptography/
> Is there any to find out how many users actually use crypto module? Can
> we use the number of hits on the documentation of sympy.crypto web page.
> If not many people are there using it then we can proceed deprecating it.
>

I'm not voting for or against. I just want to point out that these two
modules are
disjoint. They are on totally different aspects of cryptography. A third one
is here: https://github.com/jameslyons/pycipher/tree/master/pycipher

**Comment by asmeurer** [comment]
> I just want to point out that these two
modules are
disjoint. They are on totally different aspects of cryptography.

Yes, `cryptography` is a great module if you want to implement actual cryptography for a real-life application. In fact, we already recommend it at [the top of the crypto docs](https://docs.sympy.org/latest/modules/crypto.html). 

However, insomuch as the `sympy.crypto` module serves to have example implementations of cryptographic algorithms, the `cryptography` module is not a suitable replacement for it. This is in part because `cryptography` has many technical details implemented that are required to make the algorithms usable and secure. But more importantly, the core algorithms themselves are not implemented in Python in `cryptography`, but rather `cryptography` links against core libraries like openssl. 

pycipher is probably the closest to sympy.crypto from what I've seen. Are there are other high quality pure Python implementations of the sorts of things that are in `sympy.crypto` that we can point to as well?
